### ISSUE TITLE
can `types.model` be used to model generic types?

### ISSUE THREAD

**Issue by johnnyreilly** [mobxjs/mobx-state-tree#468]
This was originally [a question on StackOverflow](https://stackoverflow.com/q/46733922/761388) which didn't get an answer.  But I did get to create the `mobx-state-tree` tag ... which was nice!

I'm using `mobx-state-tree` for the first time and I'm trying to use `types.model` to model a generic type.  I'm trying to take this existing mobx code:

    export interface IAsyncProps<TResponse> {
        isRequesting: boolean;
        response?: TResponse;
        error?: string;
    }

    /**
     * Models an Async request / response - all the properties exposed are
     * decorated with `observable.ref`; i.e. they are immutable
     */
    class Async<TRequest, TResponse> implements IAsyncProps<TResponse> {
        /**
         * Whether a request is in process
         */
        @observable.ref isRequesting: boolean = false;
    
        /**
         * (optional) The response received last time a request was run
         */
        @observable.ref response?: TResponse;
    
        /**
         * (optional) The error received last time a request was run
         */
        @observable.ref error?: string;
    
        constructor(private process: (request: TRequest) => Promise<TResponse>) {
    
        }
    
        @action
        async run(request?: TRequest) {
            try {
                this.isRequesting = true;
                this.response = undefined;
                this.error = undefined;
    
                const response = await this.process(request);
                this.runSuccess(response);
            } catch (error) {
                this.runError(error);
            }
        }
    
        @action private runSuccess(response: TResponse) {
            this.response = response;
            this.isRequesting = false;
        }
    
        @action private runError(error: any) {
            this.error = error.message ? error.message : error;
            this.isRequesting = false;
        }
    
        @action
        reset() {
            this.isRequesting = false;
            this.response = undefined;
            this.error = undefined;
            return this;
        }
    }

And I want to port it across to a `types.model`.  I've got this far:

    const Async2 = types.model('Async2', {
        isRequesting: types.boolean,
        response: types.optional(/*types.? what goes here ???*/, undefined),
        error: types.optional(types.string, undefined)
    });

But I'm stuck on how I can generically type the response.  I can handle the rest of the actions - does anyone know if that's possible?

**Comment by mattiamanzati** [comment]
If you think about generics they are dynamically generated classes. So you can just write a function that given the two generics type return your model definition using them :)

Something like

Function Async(ReqType, ResType){
. Return t.model({
.  Req: ReqType,
.  Res: ResType
. })
}

I also recommend you to memoize that function, so if you call it again it will be given back the same model type without creating a new one! :)

**Comment by mattiamanzati** [comment]
Also beware that t.optional(Type, undefined) makes no sense, maybe you wanted something like
T.union(Type, t.undefined) ? :)

**Comment by johnnyreilly** [comment]
> maybe you wanted something like
T.union(Type, t.undefined)

Almost certainly :wink:

**Comment by johnnyreilly** [comment]
Okay, so I've given it a try and here's what I've come up with: (I haven't forgotten the memoisation BTW - I just want to focus on solving the initial problem first)

```
function Asyncer<TResponse>(responseType: TResponse) {
    return types.model({
        isRequesting: types.boolean,
        response: types.union(responseType /* ERROR here! */, types.undefined),
        error: types.union(types.string, types.undefined)
    });
}
```

The problem with the above code is that it errors out with:

> Argument of type 'TRequest' is not assignable to parameter of type 'IType<any, any> | ((snapshot: any) => IType<any, any>)'.
>  Type 'TRequest' is not assignable to type '(snapshot: any) => IType<any, any>'.

What this reveals to me is this: I'm trying to understand how you can apply generics to the JSON that `responseType` represents.  Not clear?  Of course!  I'm expressing myself badly.  Let me try again.

When I'm declaring a `boolean` I use `types.boolean`.
When I'm declaring a `string` I use `types.string`.

Simple.  Now when I'm declaring an object literal such as: 

```
interface SomeArbitraryStructureTheServerSent {
    name: string;
    age: number;
    favouriteWords: string[];
}
```

how do I tell `mobx-state-treej` about it?  I can't do `types.SomeArbitraryStructureTheServerSent`.   But I'd hazard a guess there is something I can do that you know about.

PS my endgame here, once I understand what I'm doing wrong, is to submit a docs PR that explains this use case as I reckon it's pretty common.

**Comment by mattiamanzati** [comment]
Ok, sorry I was from mobile and so typed a quick reply:

So, let's start by few question: Do you expect TResponse and TRequest to be mutable and defined as MST model types?
If so, you just need to adjust the generic declaration, because you need to tell TS that TResponse is a MST type.

```javascript
function Asyncer<TResponse extends IType<any, any>>(responseType: TResponse) {
    return types.model({
        isRequesting: types.boolean,
        response: types.union(responseType, types.undefined),
        error: types.union(types.string, types.undefined)
    });
}
```

or if you have existing TS interfaces, you can use something like
```javascript
function Asyncer<TResponse>(responseType: IType<any, TResponse>) {
    return types.model({
        isRequesting: types.boolean,
        response: types.union(responseType, types.undefined),
        error: types.union(types.string, types.undefined)
    });
}
```

Then you can model your response type as MST type or use types.frozen(), which is almost like "any" in TS, but will be immutable :)

**Comment by johnnyreilly** [comment]
No bother - now I'm on mobile so expect my response to be less wordy! What you've said is very helpful - `types.frozen` is almost certainly my baby as I tend to treat all my objects as immutable (if changes are required I use `Object.assign` to create a new object)

I'm away from my computer right now. I'll try and give this a go next time I'm in front of my keyboard. Thanks again and happy Sunday!

**Comment by johnnyreilly** [comment]
Okay - so I created a mobx-state-tree version which looks like this:

```
function Async2<TRequest, TResponse>(
    process: (request: TRequest) => Promise<TResponse>
) {
    return types.model('Async2', {
        isRequesting: types.optional(types.boolean, false),
        response: types.union({} as IType<any, TResponse>, types.undefined),
        error: types.union(types.string, types.undefined)
    }).actions(self => ({
        async run(request?: TRequest) {
            try {
                self.isRequesting = true;
                self.response = undefined;
                self.error = undefined;

                const response = await process(request);
                self.response = response;
                self.isRequesting = false;
            } catch (error) {
                self.error = error.message ? error.message : error;
                self.isRequesting = false;
            }
        },

        reset() {
            self.isRequesting = false;
            self.response = undefined;
            self.error = undefined;
        }
    }));
}
```

Usage looks like this:

```
// getToken: (login: ILogin) => Promise<ILoginResultView>
const loginResult2 = Async2(getToken).create();
```

However, I feel I'm definitely doing something wrong as the types are not flowing through.  What I mean by this is that when I access `response` it's type is `TResponse` and not `ILoginResultView`.

Does usage of this look correct?  Things that I am concerned about:

1. TResponse rather that the actual type
2. How I am declaring `response` up top.  When I declare the model I don't have a response yet.  That property will be `undefined` until the `run` action completes.  I've assumed that using an empty object (`{}`) is okay?

Thanks again for your help; I really appreciate it.

**Comment by mattiamanzati** [comment]
You need to provide a MST type for the response; {} is not a valid MST type, so that's why it's throwing.
You need something like this:

```
import {process} from "mobx-state-tree"

function Async2<TRequest, TResponse>(
    responseType: IType<any, TResponse>,
    process: (request: TRequest) => Promise<TResponse>
) {
    return types.model('Async2', {
        isRequesting: types.optional(types.boolean, false),
        response: types.union(responseType, types.undefined),
        error: types.union(types.string, types.undefined)
    }).actions(self => ({
        run: process(function* run(request?: TRequest) {
            try {
                self.isRequesting = true;
                self.response = undefined;
                self.error = undefined;

                const response = yield process(request);
                self.response = response;
                self.isRequesting = false;
            } catch (error) {
                self.error = error.message ? error.message : error;
                self.isRequesting = false;
            }
        }),

        reset() {
            self.isRequesting = false;
            self.response = undefined;
            self.error = undefined;
        }
    }));
}

// usage
const loginResult2 = Async2(types.frozen, getToken).create();
// or a typed one
const ResponseType = types.model({ success: types.boolean })
const loginResult2 = Async2(ResponseType, getToken).create();

```
sorry for quick reply, lot of work today here

**Comment by johnnyreilly** [comment]
Thanks - I'll give it whirl! I'll rename `process` to `flow` in my import since that's what I think it's going to end up being (from https://github.com/mobxjs/mobx-state-tree/pull/435 )

**Comment by johnnyreilly** [comment]
Okay so I did this:

```
import { types, IType, process as flow } from 'mobx-state-tree';

export function Async2<TRequest, TResponse>(
    responseType: IType<any, TResponse>,
    process: (request: TRequest) => Promise<TResponse>
) {
    return types.model('Async2', {
        isRequesting: types.optional(types.boolean, false),
        response: types.union(responseType, types.undefined),
        error: types.union(types.string, types.undefined)
    }).actions(self => ({
        run: flow(function* run(request?: TRequest) {
            try {
                self.isRequesting = true;
                self.response = undefined;
                self.error = undefined;

                const response: TResponse = yield process(request);
                self.response = response;
                self.isRequesting = false;
            } catch (error) {
                self.error = error.message ? error.message : error;
                self.isRequesting = false;
            }
        }),

        reset() {
            self.isRequesting = false;
            self.response = undefined;
            self.error = undefined;
        }
    }));
}

// usage
const loginResult = Async2(types.frozen, this.api.security.createToken).create();
```

Unfortunately usage isn't ideal; the return type ends up being:

```
(property) SecurityStore.loginResult: {
    isRequesting: boolean;
    response: TResponse;
    error: string;
} & {
    run: () => Promise<{}>;
    reset(): void;
} & {
    readonly $treenode?: any;
} & ISnapshottable<{
    isRequesting?: any;
    response?: any;
    error?: any;
}>
```

So `response` doesn't land up being the specific `TResponse` type I'm hoping for.  I've even tried this:

```
  loginResult = Async2(types.frozen as ISimpleType<ILoginResultView>, this.api.security.createToken).create();
```

Thinking that might force the type inference I was hoping for.  It doesn't work though.  

I think it's probably worth dwelling for a second on things that I'm finding awkward / weird.  Not because things necessarily need to change in mobx-state-tree.  They may well be like that for good reason.  But rather because I want to be sure that my choices are things that are genuinely "not the mobx-state-tree-way".  So here goes:

I have an `Async` mobx class. I planned to port that to be a mobx-state-tree model.  `Async` is designed to encapsulate a `Fetch`.  It contains an action / method which, given parameters, fires off a `Fetch` (well strictly speaking it invokes the process that it was originally `new-ed` up with).  It also contains state; the response **OR** error from the last time the process was invoked.  The response **OR** error are immutable; they will be replaced each time the process runs.  They are never updated and so are consequently marked with `@observable.ref`

Porting doesn't seem to be a straightforward prospect for various reasons:

- all state in mobx-state-tree has to be a `model`.  `model`s are roughly like interfaces in TypeScript in that they model data structures.  **Alas it's not possible to re-use TypeScript interfaces with `mobx-state-tree` - you need to redefine interfaces as MST models**.  

- It's not possible (or at least I think it's not possible) to just use JSON and keep type safety. `types.frozen as ISimpleType<ILoginResultView>` won't give you what you want. Also, you can't rely on type inference which is definitely my preference; I want to have types flow through the system and avoid re-typing across files when it's possible.

I'm starting to think that I'm trying to force the issue somewhat; basically I'm not going with the way that mobx-state-tree is designed to work and it's fighting me.  It's supposed to be opinionated and so I guess that's fine.  I'm taking a simple mobx app and porting to mobx-state-tree.  I think that's probably a bad idea based on my experience so far.  I think I'll probably take a look at the [bookshop example](https://github.com/mobxjs/mobx-state-tree/tree/master/packages/mst-example-bookshop) as it seems to illustrate how ajax etc would fit in with mobx-state-tree.

Please feel free to set me straight; I think there's a high probability that I'm doing something wrong here.  I'm still very much the mobx-noob! :smile:  Thanks again for taking the time to respond to me; I really appreciate it.

**Comment by mattiamanzati** [comment]
Uhm, just a quesion, what do you expect response to be? 

SecurityStore.loginResult: {
    isRequesting: boolean;
    response: TResponse;
    error: string;
} 

I see it as TResponse, so should be ok, right?

**Comment by johnnyreilly** [comment]
I expect the response to be `ILoginResultView` - which flows through from the API usage.

So with my existing mobx code I have:

```
loginResult = new Async(createToken);
```

and the resulting type of `loginResult` is:

```
loginResult: Async<ILogin, ILoginResultView>
```

So my assumption was that I could move to replace my `Async` class with an `Async` model like so:

```
  loginResult = Async2(types.frozen, createToken).create();
```

However the definition ends up as:

```
loginResult: {
    isRequesting: boolean;
    response: TResponse;
    error: string;
} & {
    run: () => Promise<{}>;
    reset(): void;
} & {
    readonly $treenode?: any;
} & ISnapshottable<{
    isRequesting?: any;
    response?: any;
    error?: any;
}>
```

i.e. `response` doesn't have the type from the API

**Comment by mattiamanzati** [comment]
Wait, if you try to get the type of "loginResult.create().response", is'nt that what you want?

**Comment by johnnyreilly** [comment]
Nope

**Comment by johnnyreilly** [comment]
```Async2(types.frozen, createToken).create().response``` has the type: `response: TResponse`

**Comment by FredyC** [comment]
It's really interesting discussion here. I am curious, did you figure it out eventually @johnnyreilly?

**Comment by mjewell** [comment]
Just hit the same problem... Here's a more minimal example of the issue:

```ts
function createMyModel<S, T>(ValueType: IType<S, T>) {
  return types.model('MyModel', {
    value: ValueType
  });
}

const MyStringModel = createMyModel(types.string);

/*
const MyStringModel: IModelType<{
    value?: any;
}, {
    value: T;                 <--- why is this T and not string?
}>
*/
```

vs 

```ts
const MyStringModel = types.model('MyModel', {
  value: types.string
});

/*
const MyStringModel: IModelType<{
    value?: any;
}, {
    value: string;
}>
*/
```

**Comment by mweststrate** [comment]
Can I suggest any further questions in this thread are accompanied by a typescript based codesandbox excerpt of the types? Reasoning about what type typescript infers from a piece of code is much much harder then just hovering to see what the type inference does :)

**Comment by mweststrate** [comment]
On a side node TS types will stay an 80% correctness things until conditional types are publicly available (should be in TS 2.8); without them we cannot type snapshots correctly.

@mjewell does it in your case help to use `const MyStringModel = createMyModel<string, string>(types.string)`?

**Comment by mjewell** [comment]
Nope, still `{ value: T }`: https://codesandbox.io/s/y069pxwxlx

**Comment by mweststrate** [comment]
Hmm interesting that this doesn't infer further. I think we should
reinvestigate this after 2.8

Op di 6 feb. 2018 om 09:00 schreef Michael Jewell <notifications@github.com

**Comment by mjewell** [comment]
Updated the code sandbox. Looks like somehow it figures out the types correctly for `views` and `actions`, in case that gives you any ideas about what's going on

**Comment by ConneXNL** [comment]
I think i just also got in this problem.

**Comment by mweststrate** [comment]
Closing as #705 should fix this.
